### DISCUSSION TITLE
Running tasks in parallel

### DISCUSSION THREAD

**Discussion by e3b0c442** [dagger/dagger#5644] [Help]
Is there a way to run tasks in parallel aside from creating child pipelines?

Basic example, I have a Node app that I'm testing. I would like to install the deps, and then run the linting and unit testing tasks in parallel, then rejoin for the production container build.

Are pipelines the ideal way to accomplish this?

Thanks!

**Comment by jpadams** [comment]
Hi @e3b0c442  👋  

Improving the DX for this situation is ongoing: https://github.com/dagger/dagger/issues/4205#issuecomment-1533918410
You can use a combination of the `Sync()` function and language-specific constructs like Go Error Groups to accomplish this today: https://github.com/dagger/dagger/issues/4766

Here's an example that might help illustrate this approach: https://github.com/jpadams/dagger-examples/blob/parallel/go/npm-build/ci.go

Notice below that the pull of the dagger engine and `node:18` images, plus the copy of the src directory, and the run of `npm install` all  took around 14s and the whole run about 50s. So we have about 36s for the `lint`, `tests`, `build`.

The `lint` had a 10s sleep, the `tests` had a 20s sleep, and the `build` had a 15s sleep. Adding those up, you'd expect 45s if serial execution.
But since we have `lint` and `tests` running in parallel in the error group, they take the 20s that the longest one takes. So 20s+15s=35s, which is about what we see.

 If we didn't want both of these to have to finish before `build`, we could leave them all in parallel, of course.
 
```
npm-build ➤ dagger run go run ci.go                                                                                   git:parallel*
╮
▽ init
█ [6.59s] connect
┣ [6.49s] starting engine
┣ [0.03s] starting session
┻
█ [50.2s] go run ci.go
┃
┃ > nodetest@1.0.0 build
┃ > echo "Build starting..." && sleep 15 && echo "...Build done!!!" && exit 0
┃
┃ Build starting...
┃ ...Build done!!!
┣─╮
│ ▽ host.directory .
│ █ [0.03s] upload .
│ ┣ [0.01s] transferring eyJvd25lcl9jbGllbnRfaWQiOiJyNWhmanhlZ3gwdTg0YzNwanRoODJ6aWRkIiwicGF0aCI6Ii4iLCJpbmNsdWRlX3BhdHRlcm5zIjpudWxsLCJleGNsdWRlX3BhdHRlcm5zIjpudWxsLCJmb2xsb3dfcGF0aHMiOm51bGwsInJlYWRfc2luZ2xlX2ZpbGVfb25seSI6ZmFsc2UsIm1heF9maWxlX3NpemUiOjB9:
│ █ [0.01s] copy .
│ ┣─╮ copy .
│ ┻ │
┣─╮ │
│ ▽ │ from node:18
│ █ │ [0.34s] resolve image config for docker.io/library/node:18
│ █ │ [5.24s] pull docker.io/library/node:18
│ ┣ │ [0.01s] resolve docker.io/library/node:18@sha256:11e9c297fc51f6f65f7d0c7c8a8581e5721f2f16de43ceff1a199fd3ef609f95
│ ┣ │ [0.15s] ████████████████ sha256:d75a90c8895c531a0d4e0ba067abef7432eae56c7b420f2caaff7d556e8e311a
│ ┣ │ [0.57s] ████████████████ sha256:715cea74ecbb15cb82efef1e77dd60c31d90b01d1286d6f39b4562afaebe75f3
│ ┣ │ [2.44s] ████████████████ sha256:6095ade0ed48a7cf654be8c76a68b547ef9a8b4a69879e7b65cea73890d5a4ca
│ ┣ │ [0.43s] ████████████████ sha256:b99a1b5ca5eb4387b99702275f25ac7be75bf137997a456a06311edc82e5d17c
│ ┣ │ [0.34s] ████████████████ sha256:796831e4446804333d6e990fa38bf6c68632606d9b2eed925761d84625ada36e
│ ┣ │ [4.78s] ████████████████ sha256:a56ae3b61eb9574588be7e73e31c31798e2cbf75f53f1f824d855afdf2be6437
│ ┣ │ [2.07s] ████████████████ sha256:003f1109a21287fa17dc866e87e8c6685113960cbb0379fee8f42b83de63c647
│ ┣ │ [1.82s] ████████████████ sha256:a014e5e7d08c37cf1703b97e701ccdc850e4a18d0ee679f03aa875dcd520aa85
│ ┻ │
█◀──╯ [0.02s] copy / /src
█ [1.32s] exec docker-entrypoint.sh npm install
┃
┃ added 57 packages, and audited 58 packages in 640ms
┃
┃ 7 packages are looking for funding
┃   run `npm fund` for details
┃
┃ found 0 vulnerabilities
█ [11.0s] exec docker-entrypoint.sh npm run lint
┃
┃ > nodetest@1.0.0 lint
┃ > echo "Linter starting..." && sleep 10 && echo "...Linter done!" && exit 0
┃
┃ Linter starting...
┃ ...Linter done!
┃ npm notice
┃ npm notice New minor version of npm available! 9.6.7 -> 9.8.1
┃ npm notice Changelog: <https://github.com/npm/cli/releases/tag/v9.8.1>
┃ npm notice Run `npm install -g npm@9.8.1` to update!
┃ npm notice
█ [21.0s] exec docker-entrypoint.sh npm run test
┃
┃ > nodetest@1.0.0 test
┃ > echo "Tests starting..." && sleep 20 && echo "...Tests done!" && exit 0
┃
┃ Tests starting...
┃ ...Tests done!
┃ npm notice
┃ npm notice New minor version of npm available! 9.6.7 -> 9.8.1
┃ npm notice Changelog: <https://github.com/npm/cli/releases/tag/v9.8.1>
┃ npm notice Run `npm install -g npm@9.8.1` to update!
┃ npm notice
█ [15.4s] exec docker-entrypoint.sh npm run build
┃
┃ > nodetest@1.0.0 build
┃ > echo "Build starting..." && sleep 15 && echo "...Build done!!!" && exit 0
┃
┃ Build starting...
┃ ...Build done!!!
┃ npm notice
┃ npm notice New minor version of npm available! 9.6.7 -> 9.8.1
┃ npm notice Changelog: <https://github.com/npm/cli/releases/tag/v9.8.1>
┃ npm notice Run `npm install -g npm@9.8.1` to update!
┃ npm notice
┻
• Cloud URL: https://dagger.cloud/runs/97c7e24b-d3f5-4b38-9f09-44dc167e7daf
• Engine: dc70499a69bf (version v0.8.4)
• Duration: 50.208s
⧗ 56.8s ✔ 32
```



There are ideas like making `Sync()` able to accept multiple lazy chains to do the "rejoin" of the DAG you mention. https://github.com/dagger/dagger/issues/5083

P.S. We answer a lot of questions like this over in our Discord server, especially in the #help channel. 
More 👀 and convos over there for sure :smiley: https://discord.gg/dagger-io

**Reply by helderco** [reply]
That's right. You should use your language's concurrency feature to run things concurrently unless what you want to accomplish can be naturally declared in the DAG itself. https://github.com/dagger/dagger/issues/5083 is just a convenience, but there are trade-offs.

**Comment by kpenfound** [comment]
👋 
The tricky part to do in Dagger today is the 'rejoin' part. Ideally we can support something like the `Sync()` like @jpadams mentioned above, but currently the best way to accomplish this is documented here under "Problem 2, Workaround 1" https://github.com/dagger/dagger/issues/4205

```go
// Both return a `dagger.Container`, full async
testResult := test()
lintResult := lint()

// Do *something* that forces a wait on both
_, err := c.Container().From("alpine").
  WithMountedDirectory("/a", testResult).
  WithMountedDirectory("/b", lintResult).
  WithExec([]string{"ls"}).
  ExitCode(ctx)
return err
```

This works because it takes advantage of the lazy execution model to make the final pipeline dependent on lint and test using `WithMountedDirectory`, where the thing you're mounting is irrelevant, it just forces the dependency.

More conversation can be found in that issue for what the ideal solution should be, as well as the issue @jpadams linked and https://github.com/dagger/dagger/issues/4766

**Reply by helderco** [reply]
I don't agree that this is the best way to accomplish what is asked since it's a workaround. There are patterns for it but not this one (e.g., building a binary for multiple platforms, adding to a dir that can be exported in the end).

In other words, even though it works, it wasn't designed for this purpose (in this case, executing arbitrary and unrelated containers). The canonical way to do it today is by using the language's concurrency feature:
- https://github.com/dagger/dagger/issues/4766

Btw that example is outdated, nowadays you'd do it like:
```diff
_, err := c.Container().
-   From("alpine").
    WithMountedDirectory("/test", test.Directory("/tmp")).  
    WithMountedDirectory("/lint", lint.Directory("/tmp")).
-   WithExec([]string{"ls"}).
-   ExitCode(ctx)
+   Sync(ctx)
```
No need for a dummy image and exec anymore, but you need a `Directory` (not `Container`) if using `WithMountedDirectory`. You can also use [`RedirectStdout`](https://pkg.go.dev/dagger.io/dagger#ContainerWithExecOpts) in `WithExec` to add the stdout to a dir and add that instead, maybe getting all the outputs in the end instead of just `Sync` above.
